hydromt/data_source/geodataframe.py

```diff
@@ -0,0 +1,223 @@
+"""Generic DataSource for GeoDataFrames."""
+
+from datetime import datetime
+from logging import Logger, getLogger
+from os.path import basename, splitext
+from typing import Any, ClassVar, Dict, List, Literal, Optional
+
+import geopandas as gpd
+from pydantic import ValidationInfo, field_validator
+from pyproj import CRS
+from pyproj.exceptions import CRSError
+from pystac import Asset as StacAsset
+from pystac import Catalog as StacCatalog
+from pystac import Item as StacItem
+from pystac import MediaType
+
+from hydromt._typing import Bbox, ErrorHandleMethod, Geom, NoDataStrategy, TotalBounds
+from hydromt.data_adapter.geodataframe import GeoDataFrameAdapter
+from hydromt.driver.geodataframe_driver import GeoDataFrameDriver
+from hydromt.driver.pyogrio_driver import PyogrioDriver
+
+from .data_source import DataSource
+
+logger: Logger = getLogger(__name__)
+
+# placeholder for proper plugin behaviour later on.
+_KNOWN_DRIVERS: Dict[str, GeoDataFrameDriver] = {"pyogrio": PyogrioDriver}
+
+
+def driver_from_str(driver_str: str, **driver_kwargs) -> GeoDataFrameDriver:
+    """Construct GeoDataFrame driver."""
+    if driver_str not in _KNOWN_DRIVERS.keys():
+        raise ValueError(
+            f"driver {driver_str} not in known GeoDataFrameDrivers: {_KNOWN_DRIVERS.keys()}"
+        )
+
+    return _KNOWN_DRIVERS[driver_str](**driver_kwargs)
+
+
+class GeoDataFrameSource(DataSource):
+    """
+    DataSource for GeoDataFrames.
+
+    Reads and validates DataCatalog entries.
+    """
+
+    data_type: ClassVar[Literal["GeoDataFrame"]] = "GeoDataFrame"
+    driver: GeoDataFrameDriver
+    data_adapter: GeoDataFrameAdapter
+
+    @field_validator("driver", mode="before")
+    @classmethod
+    def _check_geodataframe_drivers(cls, v: Any, info: ValidationInfo) -> str:
+        if isinstance(v, str):
+            if v not in _KNOWN_DRIVERS:
+                raise ValueError(f"unknown driver '{v}'")
+            return driver_from_str(v, **info.data.get("driver_kwargs"))
```

Comment:
When debugging I found that `data` is still complete in in `DataSource._init_subclass_data_type` but only contains `"name"` and `"uri"` as property of the `info` object at this line.. ? It's completely magic to me what happens in between..

Reply:
Just investigated. I have this `_driver_from_str` function that now controls creating drivers, but as the `MetaDataResolver` is now injected into the driver class, just supplying this + driver kwargs does not fit the bill anymore.
Some DataCatalog entry like this would be needed? Which we can just supply as a dict and test. metadatar_esolver could be optional.
```yaml
cooldata:
  uri: "path/to/cooldata.fgb"
  driver:
    name: "pyogrio"
    metadata_resolver:
      name: "convention"
      ...metadata_resolver_kwargs
    ...driver_kwargs
  harmonization_settings:  # find a better name for this
    rename:
      cool: cold
      warm: hot
  meta:
    ...metadata
```

Reply:
A user does not define which metadata_resolver to set for a driver right? That's fixed in the code. Also its arguments are all runtime arguments when a users requests data for a specific region. Could we solve this by removing the metadata_resolver from the initialization of the driver and fixing it as a property instead?

Reply:
Ah this was not yet clear from the refinement then, in my head a `Driver` *has* a `MetaDataResolver` (and can validate if it wants), so we do need to make many combinations of drivers and metadataresolvers. What I did in the new version is setting a default resolver.

Reply:
My idea was inspired by: https://refactoring.guru/design-patterns/bridge

Reply:
My code snippets above still don't work. But those should work right? Not sure from your replies.

I thought that we did discuss during the refinement that we will may need to implement multiple drivers to accommodate for combinations of `MetaDataResolver` and data readers. We won't be able to support all combinations anyway, so for the few combinations I think it make sense to make multiple `Driver` classes.

Reply:
For this metadata resolver, I am still afraid of an explosion of driver classes and overloading the `Driver` class with too much logic. I see us writing a different Driver per dataset. Reading a certain file format to python memory representation would be the main use for a Driver. Now we also add merging multiple objects coming from the metadata resolver and adding preprocessing. If we also add finding all relevant URIs based on different convention, we are back at `DataAdapter` minus a few lines about renaming and multiplying units and the whole class becomes hard to test again. What should a `Driver` do?
Would a default `MetaDataResolver` (or `UriResolver`) not resolve most of your worries? You could give a list of resolvers to your `Driver` to validate the combination.
If not then back to the drawing board? Maybe we have to look at it through a different lens, like having different layers of `Filters` to go from a single `uri` in the `DataCatalog` to a single `GeoDataFrame` in a declarative (yaml) way, which is the goal in my opinion.

Reply:
Sorry, I should have looked at the test better. I now noticed that the syntax has changed from `driver="pyogrio"` to `driver={"name": "pyogrio"}`. That should work.

Regarding the default `MetaDataResolver`. Let's see how this works. In general I like the idea, I just wonder if we can ensure compatibility between all `MetaDataResolver` and `DataDriver`. For instance a `GeotiffDriver` could get uris of different variables that should be stacked or uris of different tiles that should be mosaiced. I think both are easier dealt with in different Drivers. But I'm happy to test your approach further!

Reply:
I will make sure that driver kwargs come at the same level, .e.g.
```python
driver={
    "name": "pyogrio",
    "kwarg1": 1,
    "kwarg2": "bar",
}
```

Reply:
> idem for
> 
> ```python
> DataSource(
>     **{
>         "name": "test",
>         "uri": r"some_uri.zarr",
>         "driver": "zarr",
>         "driver_kwargs": {},
>         "data_type": "RasterDataset",
>     }
> )
> ```
> 
> ```python
> DataSource(
>     **{
>         "name": "test",
>         "uri": r"points.geojson",
>         "driver": "pyogrio",
>         "driver_kwargs": {},
>         "data_type": "GeoDataFrame",
>     }
> )
> ```

This is not possible, as it is not possible to return a different class from the `__init__` method, as the class is already initialized.
What we can do, is to use `DataSource.model_validate`, which is standard in pydantic, or start using factories. There we can put the logic also for keeping the `unit_add` etc.